### PR TITLE
[Interop][SwiftToCxx] Support enum creation from Cxx

### PR DESCRIPTION
**Pull request by WANGJIEKE**
This pull request adds support for Swift enum creation from C++:

- Add function in `cxx_synthesis` to print `swift` namespace in C++.
- Update `DeclAndTypePrinter` to print functions needed for enum creation.
  - Fix a bug when printing cases for empty resilient enum
- Forward declare types used in Swift enums
- Add a function in `DeclAndTypeClangFuncPrinter` to print a custom generated function in C++
- Update and merge test cases

### CODE REVIEW

**Comment by WANGJIEKE** [comment]
@hyp Could you please test and review this pull request? Thanks.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test macOS platform

**Comment by hyp** [comment]
@swift-ci please test source compatibility

File: test/Interop/SwiftToCxx/enums/swift-enum-implementation-execution.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    case E::foobar:
+        assert(e.isFoobar());
+        break;
+    }
+}
+
+int main() {
+    {
+        auto e = E::x(3.14);
+        switchTest(e);
```
Comment (hyp) [inline test/Interop/SwiftToCxx/enums/swift-enum-implementation-execution.cpp:49]: we're asserting in each case, but we should also verify that we actually went through the right switch case. How about returning some sentinel value from `switchTest` e.g. a string, and asserting here that it's expected, e.g.

```
    case E::x:
         assert(e.isX());
         assert(e.getX() == 3.14);
        return "x";

...

         auto e = E::x(3.14);
         auto res = switchTest(e);
         assert(res == "e");
```

File: test/Interop/SwiftToCxx/enums/swift-enum-implementation.swift [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+// CHECK-NEXT:     alignas(E) unsigned char buffer[sizeof(E)];
+// CHECK-NEXT:     auto *thisCopy = new(buffer) E(*this);
+// CHECK-NEXT:     char * _Nonnull payloadFromDestruction = thisCopy->_destructiveProjectEnumData();
+// CHECK-NEXT:     void * _Nonnull result;
+// CHECK-NEXT:     memcpy(&result, payloadFromDestruction, sizeof(result));
+// CHECK-NEXT:     return result;
+// CHECK-NEXT:   }
+// CHECK-NEXT:   inline E E::_impl_foobar::operator()() const {
+// CHECK-NEXT:     auto result = E::_make();
+// CHECK-NEXT:     result._destructiveInjectEnumTag(5);
```
Comment (hyp) [inline test/Interop/SwiftToCxx/enums/swift-enum-implementation.swift:258]: Can you also add a test for resilient enum where we call `_destructiveInjectEnumTag` using extern "C" tag values as well?

File: test/Interop/SwiftToCxx/enums/swift-enum-implementation.swift [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+// CHECK-NEXT:     return enumVWTable->getEnumTag(_getOpaquePointer(), metadata._0);
+// CHECK-NEXT:   }
+// CHECK-NEXT:   using _impl_x = decltype(x);
+// CHECK-NEXT:   using _impl_y = decltype(y);
+// CHECK-NEXT:   using _impl_z = decltype(z);
+// CHECK-NEXT:   using _impl_w = decltype(w);
+// CHECK-NEXT:   using _impl_auto = decltype(auto_);
+// CHECK-NEXT:   using _impl_foobar = decltype(foobar);
+// CHECK:      };
+// CHECK-NEXT: decltype(E::x) E::x;
```
Comment (hyp) [inline test/Interop/SwiftToCxx/enums/swift-enum-implementation.swift:142]: can you make supporting structs emitted as `inline const ` variables? that way they're actually unique to each TU and won't conflict at link time.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test linux platform

**Review by hyp — APPROVED** [review]
LGTM

**Comment by hyp** [comment]
@swift-ci please test linux platform

**Comment by hyp** [comment]
@swift-ci please test windows platform

**Comment by hyp** [comment]
@swift-ci please smoke test macos platform

**Comment by hyp** [comment]
@swift-ci please test macos platform

**Pull request merged**